core/src/main/java/io/grpc/Deadline.java

```diff
@@ -0,0 +1,136 @@
+ *    * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ *    * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc;
+
+import com.google.common.base.Preconditions;
+
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+
+/**
+ * An absolute deadline in system time.
+ */
+public class Deadline {
+
+  private static final AtomicIntegerFieldUpdater<Deadline> expiredUpdater =
+      AtomicIntegerFieldUpdater.newUpdater(Deadline.class, "expired");
+
+  /**
+   * Create a deadline that will expire at the specified offset from the current system clock.
+   * @param duration A non-negative duration.
+   * @param units The time unit for the duration.
+   * @return A new deadline.
+   */
+  public static Deadline after(long duration, TimeUnit units) {
+    Preconditions.checkNotNull(units);
+    if (duration > 0) {
+      return new Deadline(System.nanoTime(), units.toNanos(duration), false);
+    } else {
+      // Deadlines at the current instant or earlier are already expired.
+      return new Deadline(System.nanoTime(), units.toNanos(duration), true);
+    }
+  }
+
+  private final long deadlineNanos;
+
+  @SuppressWarnings("unused")
+  private volatile int expired;
+
+  private Deadline(long baseInstant, long offset, boolean baseInstantAlreadyExpired) {
+    if (offset > 0 && Long.MAX_VALUE - offset < baseInstant) {
+      deadlineNanos = Long.MAX_VALUE;
+      expired = 0;
+    } else if (offset < 0 && Long.MIN_VALUE - offset > baseInstant) {
+      deadlineNanos = Long.MIN_VALUE;
+      expired = 1;
+    } else {
+      deadlineNanos = baseInstant + offset;
+      expired = baseInstantAlreadyExpired && offset <= 0 ? 1 : 0;
```

Comment:
Checking `offset` here seems redundant. Either check it in `after()` or check it here, but not both.